intermediate_source/parametrizations.py

```diff
@@ -0,0 +1,353 @@
+# Composing Parametrizations
+# --------------------------
+# Concatenating two parametrizations is as easy as registering them on the same tensor.
+# We may use this to create complex parametrizations from simple ones. For example, the
+# `Cayley map <https://en.wikipedia.org/wiki/Cayley_transform#Matrix_map>`_
+# maps the skew-symmetric matrices to the orthogonal matrices of positive determinant. We can
+# concatenate ``Skew`` and a parametrization that implements the Cayley map to get a layer with
+# orthogonal weight
+class CayleyMap(nn.Module):
+    def __init__(self, n):
+        self.register_buffer("Id", torch.eye(n))
+
+    def forward(self, X):
+        # (I + X)(I - X)^{-1}
+        return torch.solve(self.Id + X, self.Id - X).solution
+
+layer = nn.Linear(3, 3)
+P.register_parametrization(layer, "weight", Skew())
+P.register_parametrization(layer, "weight", CayleyMap(3))
+X = layer.weight
+assert torch.allclose(X.T @ X, torch.eye(3))  # X is orthogonal
+
+###############################################################################
+# This may also be used to prune a parametrized module, or to reuse parametrizations. For example,
+# we may use the fact that the exponential of matrices maps the symmetric matrices to the
+# Symmetric Positive Definite (SPD) matrices, and the skew-symmetric matrices to the orthogonal
+# matrices. Using these two facts, we may reuse the parametrizations
+class MatrixExponential(nn.Module):
+    def forward(X):
+        return torch.matrix_exp(X)
+
+layer_orthogonal = nn.Linear(3, 3)
+P.register_parametrization(layer_orthogonal, "weight", Skew())
+P.register_parametrization(layer_orthogonal, "weight", MatrixExponential())
+X = layer_orthogonal.weight
+assert torch.allclose(X.T @ X, torch.eye(3))  # X is orthogonal
+
+layer_spd = nn.Linear(3, 3)
+P.register_parametrization(layer_spd, "weight", Symmetric())
+P.register_parametrization(layer_spd, "weight", MatrixExponential())
+X = layer_spd.weight
+assert torch.allclose(X, X.T)                    # X is symmetric
+assert (torch.symeig(X).eigenvalues > 0.).all()  # X is positive definite
+
+###############################################################################
+# Intializing Parametrizations
+# ----------------------------
+# Parametrizations come with a mechanism to initialize them. If we implement a method
+# ``right_inverse`` with signature
+#
+# .. code-block:: python
+#
+#     def right_inverse(self, X: Tensor) -> Tensor
+#
+# it will be used when assigning to the parametrized tensor.
+#
+# Let's upgrade our implementation of the ``Skew`` class to support this
+class Skew(nn.Module):
+    def forward(self, X):
+        A = X.triu(1)
+        return A - A.transpose(-1, -2)
+
+    def is_skew(self, X):
+        return torch.allclose(X, -X.transpose(-1, -2))
+
+    def right_inverse(self, A):
+        if not self.is_skew(A):
+            raise ValueError(f"The provided matrix {A} is not skew-symmetric")
+        return A.triu(1)
```

Comment:
Should we mention here that we only get the upper elements because these are the elements used in the forward?